java/apirag-starter/src/main/java/com/datasqrl/ai/LowLevelAgent.java

```diff
@@ -0,0 +1,155 @@
+          "temperature", 0.3)));
+
+  /**
+   * Initializes a command line chat bot using OpenAI
+   *
+   * @param toolsPath     A path to a tools file in JSON format
+   * @param apiExecutor   An APIExecutor for the tools
+   * @param systemPrompt  The system message for the LLM
+   * @param openAIKey     The OpenAI API key to call the API
+   */
+
+  public LowLevelAgent(Path toolsPath, APIExecutor apiExecutor, String systemPrompt, String openAIKey) throws IOException {
+    this.systemPrompt = systemPrompt;
+    this.service = new OpenAiService(openAIKey, Duration.ofSeconds(60));
+    List<RuntimeFunctionDefinition> tools = ToolsBackendFactory.readTools(toolsPath);
+    this.toolsBackend = ToolsBackendFactory.of(tools, Map.of(APIExecutorFactory.DEFAULT_NAME, apiExecutor));
+  }
+
+  /**
+   * Starts the chatbot on the command line which will accepts questions and produce responses.
+   * Type "exit" to terminate.
+   */
+  public void start(Map<String, Object> context) throws IOException {
+    Scanner scanner = new Scanner(System.in);
+    OpenAIModelBindings modelBindings = new OpenAIModelBindings(chatConfig);
+    ChatSession<ChatMessage, ChatFunctionCall> session = new ChatSession<>(toolsBackend, context, systemPrompt, modelBindings);
+
+    System.out.print("First Query: ");
+    ChatMessage firstMsg = new UserMessage(scanner.nextLine());
+    session.addMessage(firstMsg);
+
+    while (true) {
+      ContextWindow<ChatMessage> contextWindow = session.getContextWindow();
+      ChatCompletionRequest chatCompletionRequest = ChatCompletionRequest
+          .builder()
+          .model(chatConfig.getModelName())
+          .messages(contextWindow.getMessages())
+          .functions(contextWindow.getFunctions())
+          .functionCall("auto")
+          .n(1)
+          .topP(chatConfig.getTopP())
+          .temperature(chatConfig.getTemperature())
+          .maxTokens(chatConfig.getMaxOutputTokens())
+          .logitBias(new HashMap<>())
+          .build();
+      Flowable<ChatCompletionChunk> flowable = service.streamChatCompletion(chatCompletionRequest);
+
+      AtomicBoolean isFirst = new AtomicBoolean(true);
+      AssistantMessage responseMessage = service.mapStreamToAccumulator(flowable)
+          .doOnNext(accumulator -> {
+            if (accumulator.isFunctionCall()) {
+              if (isFirst.getAndSet(false)) {
+                System.out.println("Executing function " + accumulator.getAccumulatedChatFunctionCall().getName() + "...");
+              }
+            } else {
+              if (isFirst.getAndSet(false)) {
+                System.out.print("Response: ");
+              }
+              if (accumulator.getMessageChunk().getContent() != null) {
+                System.out.print(accumulator.getMessageChunk().getContent());
+              }
+            }
+          })
+          .doOnComplete(System.out::println)
+          .lastElement()
+          .blockingGet()
+          .getAccumulatedMessage();
```

Comment:
@mbroecheler This part of the code makes it complicated. I don't even understand what is happening here. Do you think we should keep this?

Reply:
Yes, I like it ;-)